text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+   `didCreate` and `didUpdate` hooks, and after any child element's
+   `{{on-insert}}` modifiers, preserving the semantics of `didInsertElement`.
+
+   `{{on-insert}}` receives a function with the following signature:
+
+   ```ts
+   type OnInsertHandler = function(element: Element): void;
+   ```
+
+2. `{{on-destroy}}`, which runs immediately before the element is removed.
+   `{{on-destroy}}` is guaranteed to run before the enclosing component's
+   `didUpdate` and `destroy` hooks, and after and child element's
+   `{{on-destroy}}` modifiers, preserving the semantics of `willDestroyElement`.
+
+   `{{on-destroy}}` receives a function with the following signature:
+
+   ```ts
+   type OnDestroyHandler = function(element: Element): void;
+   ```
+
+#### Default Binding of Functions
+
+Currently, the `{{action}}` modifier is the only modifier which ships with
+Ember, and `{{action}}` by default binds functions to the context of the
+_component_. With nested values, this behavior does not particularly make sense:
+
+```hbs
+<div {{on-insert this.animationService.registerElement}}></div>
+```
+
+In this example, `on-insert` would effectively be running
+`this.animationService.registerElement.call(this, element)`, where `this` refers
+to the _component_ and not the _animation service_.
+
+The `bind` RFC, which runs parallel to this one, proposes that a new helper with
+the opposite semantics be introduced. If the `bind` RFC is accepted before this
+RFC is implemented, then the default behavior when using `on-insert` should
+match with the `bind` helper, and _not_ the `action` helper.
+
+### Untracked Properties
+
+Currently in Ember components, all properties are implicitly watched if they are
+bound from a template, and can be updated using Ember's `set` function:
+
+```hbs
+{{this.count}}
+
+<button onclick={{action this.incrementCount}}>
+  Increment
+</button>
+```
+```js
+export default Component.extend({
+  count: 0,
+
+  incrementCount() {
+    this.set('count', this.count + 1);
+  }
+});
+```
+
+By contrast, Glimmer components will require that users opt-in to change
+tracking for all values in the component. The following code would _not_ update:
+
+```hbs
+{{this.count}}
+
+<button onclick={{action this.incrementCount}}>
+  Increment
+</button>
+```
+```js
+export default class CounterComponent extends GlimmerComponent {
+  count = 0;
+
+  incrementCount() {
+    this.set('count', this.count + 1);
+  }
+}
+```
+
+In order to get this code working, users would have to explicitly mark this code
+as tracked. There are 2 ways to do this:
+
+1. Turn `count` into a computed property
+2. Turn count into a tracked property (as proposed in the Tracked Properties RFC)
+
+With computed properties, this would look like:
+
+```js
+export default class CounterComponent extends GlimmerComponent {
+  @computed('_count')
+  get count() {
+    return this._count;
+  }
+
+  incrementCount() {
+    this.set('_count', this.count + 1);
```

Comment:
semi orthogonal, but I suspect the tracked RFC (or maybe some short/medium term effort) may want to promote computed properties to also having a setter, that way `this.set` would not need to be used in this scenario. This will likely easy transition pains, especially for addon authors.